### PR TITLE
netplan used to configure main Foreman interface

### PR DESCRIPTION
**Pull request by pacaj2am**
in newer Ubuntus netplan.io is used instead of /etc/network/interfaces. I created small adjustments of the preseed network configuration to accomodate the change.

The main difference to the yaml file generated by preseed itsel if that if you have a secondary DNS server, it is added to DNS section of netplan.

### CODE REVIEW

File: provisioning_templates/snippet/preseed_networking_setup.erb [hunk 1]
```diff
@@ -7,12 +7,37 @@ description: this will configure your host networking, it configures your primar
...
+    #<%= @host.primary_interface.identifier %>
+    $real:
+<% if host_dhcp -%>
+      dhcp4: true
+<% else -%>
+      addresses: [ <%= @host.ip %>/<%= host_subnet.cidr %><% if @host.ip6 && host_subnet6 && !host_dhcp6 -%>, <%= @host.ip6 %>/<%= host_subnet6.cidr %><% end -%> ]
+      gateway4: <%= host_subnet.gateway %>
+      nameservers:
+          search: [ <%= @host.domain %> ]
+          addresses: [ <%= host_subnet.dns_primary %><% if host_subnet.dns_secondary -%>, <%= host_subnet.dns_secondary %><% end -%> ]
```
Comment (ekohl) [inline provisioning_templates/snippet/preseed_networking_setup.erb:40]: I feel like we repeat this code a lot. We should really add a helper that presents the DNS servers as a list

File: provisioning_templates/snippet/preseed_networking_setup.erb [hunk 2]
```diff
@@ -7,12 +7,37 @@ description: this will configure your host networking, it configures your primar
...
+  version: 2
+  renderer: networkd
+  ethernets:
+    #<%= @host.primary_interface.identifier %>
+    $real:
+<% if host_dhcp -%>
+      dhcp4: true
+<% else -%>
+      addresses: [ <%= @host.ip %>/<%= host_subnet.cidr %><% if @host.ip6 && host_subnet6 && !host_dhcp6 -%>, <%= @host.ip6 %>/<%= host_subnet6.cidr %><% end -%> ]
+      gateway4: <%= host_subnet.gateway %>
```
Comment (ekohl) [inline provisioning_templates/snippet/preseed_networking_setup.erb:36]: We also have an (optional) IPv6 gateway. I don't know netplan.io, but this format suggests we need a `gateway6` as well.

File: provisioning_templates/snippet/preseed_networking_setup.erb [hunk 3]
```diff
@@ -7,12 +7,37 @@ description: this will configure your host networking, it configures your primar
...
+cat << EOF > /etc/netplan/01-netcfg.yaml
+# This file describes the network interfaces available on your system
+# For more information, see netplan(5).
+network:
+  version: 2
+  renderer: networkd
+  ethernets:
+    #<%= @host.primary_interface.identifier %>
+    $real:
+<% if host_dhcp -%>
```
Comment (ekohl) [inline provisioning_templates/snippet/preseed_networking_setup.erb:32]: There's also `host_dhcp6` which probably needs a similar entry

File: provisioning_templates/snippet/preseed_networking_setup.erb [hunk 4]
```diff
@@ -71,3 +96,4 @@ iface $real inet6 static
 <% end -%>
 EOF
 <% end -%>
+<% end -%>
```
Comment (ekohl) [inline provisioning_templates/snippet/preseed_networking_setup.erb]: nit: could you ensure a newline here?

File: provisioning_templates/snippet/preseed_networking_setup.erb [hunk 5]
```diff
@@ -7,12 +7,37 @@ description: this will configure your host networking, it configures your primar
...
+<% is_ubuntu = @host.operatingsystem.name == 'Ubuntu' -%>
+<% os_major = @host.operatingsystem.major.to_i -%>
 <% host_subnet = @host.subnet -%>
 <% host_dhcp = host_subnet.nil? ? false : host_subnet.dhcp_boot_mode? -%>
 <% host_subnet6 = @host.subnet6 -%>
 <% host_dhcp6 = host_subnet6.nil? ? false : host_subnet6.dhcp_boot_mode? -%>
 
 real=`ip -o link | awk '/<%= @host.mac -%>/ {print $2;}' | sed s/://`
+<%# Ubuntu 18.04 and newer uses netplan instead of /etc/network/interfaces -%>
+<% if is_ubuntu && os_major > 17 -%>
```
Comment (ekohl) [inline provisioning_templates/snippet/preseed_networking_setup.erb]: Nit: I prefer `>= 18`

**Review by ekohl — COMMENTED** [review]
Looks good, even though I don't know netplan.

File: provisioning_templates/snippet/preseed_networking_setup.erb (hunk 3 above)
Comment (pacaj2am) [reply provisioning_templates/snippet/preseed_networking_setup.erb:32]:
```suggestion
<% if host_dhcp6 -%>
       dhcp6: true
 <% end -%>
 <% if host_dhcp -%>
```
added

File: provisioning_templates/snippet/preseed_networking_setup.erb (hunk 2 above)
Comment (pacaj2am) [reply provisioning_templates/snippet/preseed_networking_setup.erb:36]:
```suggestion
      gateway4: <%= host_subnet.gateway %>
       <% if host_subnet6.gateway -%>gateway6: <%= host_subnet6.gateway %><% end -%>
```
added

File: provisioning_templates/snippet/preseed_networking_setup.erb (hunk 1 above)
Comment (pacaj2am) [reply provisioning_templates/snippet/preseed_networking_setup.erb:40]: sorry, but I am not ruby dev, I can not help wit that much

File: provisioning_templates/snippet/preseed_networking_setup.erb (hunk 4 above)
Comment (pacaj2am) [reply provisioning_templates/snippet/preseed_networking_setup.erb]: added

File: provisioning_templates/snippet/preseed_networking_setup.erb (hunk 5 above)
Comment (pacaj2am) [reply provisioning_templates/snippet/preseed_networking_setup.erb]:
```suggestion
<% if is_ubuntu && os_major >= 18 -%>
```

changed as well

File: provisioning_templates/snippet/preseed_networking_setup.erb (hunk 3 above)
Comment (pacaj2am) [reply provisioning_templates/snippet/preseed_networking_setup.erb:32]: one should change that for /etc/network/interfaces aswell - it is not tehre - didn't know wheter that was intentional

**Comment by pacaj2am** [comment]
there is quite poor documentation of netplan, some examples are here: https://netplan.io/examples

File: provisioning_templates/snippet/preseed_networking_setup.erb (hunk 1 above)
Comment (ekohl) [reply provisioning_templates/snippet/preseed_networking_setup.erb:40]: This was more of a general "it'd be good to have" remark than as a request for it here. I can see how it came across like that though :)

File: provisioning_templates/snippet/preseed_networking_setup.erb (hunk 3 above)
Comment (ekohl) [reply provisioning_templates/snippet/preseed_networking_setup.erb:32]: Maybe the old interfaces assumes that if you select DHCP it'll automatically do that. Perhaps it overlaps with IPv6 stateless configuration and we don't have a good way to distinguish those two.

**Comment by pacaj2am** [comment]
Hi, how can I help to finish the review or with testing?

File: provisioning_templates/snippet/preseed_networking_setup.erb [hunk 6]
```diff
@@ -7,12 +7,41 @@ description: this will configure your host networking, it configures your primar
...
+  ethernets:
+    #<%= @host.primary_interface.identifier %>
+    $real:
+<% if host_dhcp6 -%>
+      dhcp6: true
+<% end -%>
+<% if host_dhcp -%>
+      dhcp4: true
+<% else -%>
+      addresses: [ <%= @host.ip %>/<%= host_subnet.cidr %><% if @host.ip6 && host_subnet6 && !host_dhcp6 -%>, <%= @host.ip6 %>/<%= host_subnet6.cidr %><% end -%> ]
```
Comment (ekohl) [inline provisioning_templates/snippet/preseed_networking_setup.erb:35]: We have some users who use IPv6 only. I'd advise on using multiple lines YAML (for both v4 and v6) and each with their own if statements. Probably easier than to determine when to add a comma or not.

File: provisioning_templates/snippet/preseed_networking_setup.erb [hunk 7]
```diff
@@ -7,12 +7,41 @@ description: this will configure your host networking, it configures your primar
...
+      gateway4: <%= host_subnet.gateway %>
+      <% if host_subnet6.gateway -%>gateway6: <%= host_subnet6.gateway %><% end -%>
```
Comment (ekohl) [inline provisioning_templates/snippet/preseed_networking_setup.erb:37]: Won't this need a newline?

File: provisioning_templates/snippet/preseed_networking_setup.erb [hunk 8]
```diff
@@ -7,12 +7,41 @@ description: this will configure your host networking, it configures your primar
...
 
 real=`ip -o link | awk '/<%= @host.mac -%>/ {print $2;}' | sed s/://`
+<%# Ubuntu 18.04 and newer uses netplan instead of /etc/network/interfaces -%>
+<% if is_ubuntu && os_major >= 18 -%>
+cat << EOF > /etc/netplan/01-netcfg.yaml
+# This file describes the network interfaces available on your system
+# For more information, see netplan(5).
+network:
+  version: 2
+  renderer: networkd
```
Comment (lzap) [inline provisioning_templates/snippet/preseed_networking_setup.erb:25]: Is this sane default for Ubuntu? I mean, in Fedora world the network daemon of choice is NetworkManager.

**Comment by lzap** [comment]
It would probably make sense to extract the netplan.io part into a separate snippet which could be reused in other templates later on.

File: provisioning_templates/snippet/preseed_networking_setup.erb (hunk 8 above)
Comment (ekohl) [reply provisioning_templates/snippet/preseed_networking_setup.erb:25]: https://blog.ubuntu.com/2017/07/05/quick-and-easy-network-configuration-with-netplan suggests networkd is the default but I don't know if that's the final version.

Comment (lzap) [reply provisioning_templates/snippet/preseed_networking_setup.erb:25]: They elaborate on some wifi limitation tho, can you please make this a host parameter? Use any default you want, but user should be able to redefine this.

**Comment by stefanlasiewski** [comment]
Has anyone worked on a template which supports more than one interface? The format would look something like this:

```
network:
  version: 2
  renderer: networkd
  ethernets:
    ens160:
      dhcp4: yes
    ens192:
      addresses:
      - 172.16.100.100/24
      gateway4: 172.16.100.1
      nameservers:
        search: [example.org]
        addresses: [172.16.100.200,8.8.8.8,4.4.4.4]
      dhcp4: false
```

File: provisioning_templates/snippet/preseed_networking_setup.erb (hunk 8 above)
Comment (stefanlasiewski) [reply provisioning_templates/snippet/preseed_networking_setup.erb:25]: I believe that `networkd` is used as the default renderer, and specifying it in this file is not necessary. 

The 18.04 manpage at https://manpages.ubuntu.com/manpages/bionic/man5/netplan.5.html says the following, and this should be the official word:

>        · Configured  devices get handled by systemd-networkd by default, unless explicitly marked
>         as managed by a specific renderer (NetworkManager)

**Comment by stefanlasiewski** [comment]
I was trying to figure out how to submit my own fixes to this PR, but I can't figure out how to do that git-wise.

I also took a different approach, and submitted my own PR with credit to pacaj2am.

**Review by stefanlasiewski — CHANGES_REQUESTED** [review]
Looks like you accidentally commited some Visual Studio binaries?

**Comment by lzap** [comment]
This needs more work, but please keep working!

**Comment by stefanlasiewski** [comment]
I plan to get back to this after getting through a busy work period.

-= Stefan

**Comment by tbrisker** [comment]
This PR has been inactive for a long time now, closing. Feel free to reopen it if it is still relevant. Thanks @pacaj2am !

**Pull request closed without merging**